components/script/dom/console.rs

```diff
@@ -55,14 +64,138 @@ where
+                value: HandleValue,
+                mut parents: Vec<u64>,
+            ) -> DOMString {
+                if parents.len() >= MAX_LOG_DEPTH {
+                    return DOMString::from("...");
+                }
+                let value_bits = value.asBits_;
+                if parents.contains(&value_bits) {
+                    return DOMString::from("[circular]");
+                }
+                parents.push(value_bits);
+                if value.is_undefined() {
+                    // This produces a better value than "(void 0)" from JS_ValueToSource.
+                    DOMString::from("undefined")
+                } else if value.is_object() {
+                    rooted!(in(cx) let mut obj = value.to_object());
+                    let mut object_class = ESClass::Other;
+                    if !GetBuiltinClass(cx, obj.handle().into(), &mut object_class as *mut _) {
+                        return DOMString::from("/* invalid */");
+                    }
+                    if object_class != ESClass::Array && object_class != ESClass::Object {
+                        return value_to_source(cx, value);
+                    }
+
+                    let mut ids = IdVector::new(cx);
+                    if !GetPropertyKeys(
+                        cx,
+                        obj.handle().into(),
+                        jsapi::JSITER_OWNONLY | jsapi::JSITER_SYMBOLS,
+                        ids.handle_mut(),
+                    ) {
+                        return DOMString::from("/* invalid */");
+                    }
+                    let mut explicit_keys = object_class == ESClass::Object;
+                    let mut props = Vec::with_capacity(ids.len());
+                    for id in &*ids {
+                        rooted!(in(cx) let id = *id);
+                        rooted!(in(cx) let mut desc = PropertyDescriptor::default());
+
+                        let mut is_none = false;
+                        if !JS_GetOwnPropertyDescriptorById(
+                            cx,
+                            obj.handle().into(),
+                            id.handle().into(),
+                            desc.handle_mut().into(),
+                            &mut is_none,
+                        ) {
+                            return DOMString::from("/* invalid */");
+                        }
+
+                        rooted!(in(cx) let mut property = UndefinedValue());
+                        if !JS_GetPropertyById(cx, obj.handle(), id.handle(), property.handle_mut())
+                        {
+                            return DOMString::from("/* invalid */");
+                        }
+
+                        if !explicit_keys {
+                            if id.is_int() {
+                                if let Ok(id_int) = usize::try_from(id.to_int()) {
+                                    explicit_keys = props.len() != id_int;
+                                } else {
```

Comment:
Can you just do `to_int()` here without first checking whether this is an integer?

Reply:
`to_int` asserts that it's called on an integer - skipping the check would cause panics on non-integer keys (the `usize::try_from` call is returning an Option, not `to_int`).